README.md
# top [![Build Status](https://secure.travis-ci.org/fengmk2/top.png)](http://travis-ci.org/fengmk2/top) [![Coverage Status](https://coveralls.io/repos/fengmk2/top/badge.png)](https://coveralls.io/r/fengmk2/top)

[![NPM](https://nodei.co/npm/top.png?downloads=true&stars=true)](https://nodei.co/npm/top/)

[Taobao Open API](http://open.taobao.com/) Client.

## Install

```bash
$ npm install top
```

## Usage

```js
var top = require('top');
var client = top.createClient({
  appkey: 'your key',
  appsecret: 'your secret'
});
// invoke 'taobao.user.get': http://api.taobao.com/apidoc/api.htm?path=cid:1-apiId:1
client.taobao_user_get({nick: 'qleelulu', fields: 'user_id,nick,seller_credit'}, function (err, user) {
  console.log(user);
});
```

## APIs

You can find all the [Taobao open APIs](http://open.taobao.com/doc/api_list.htm?id=102).

* User
 * taobao.user.buyer.get 查询买家信息API
 * taobao.user.get 获取单个用户信息
 * taobao.user.seller.get 查询卖家用户信息
 * taobao.users.get 获取多个用户信息
* Category (Working)
 * taobao.itemcats.authorize.get 查询商家被授权品牌列表和类目列表
 * taobao.itemcats.get 获取后台供卖家发布商品的标准商品类目
 * taobao.itemcats.increment.get 增量获取后台类目数据
 * taobao.itemprops.get 获取标准商品类目属性
 * taobao.itempropvalues.get 获取标准类目属性值
 * taobao.topats.itemcats.get 全量获取后台类目数据
* Shop
 * taobao.shop.get (白名单用户才能调用)

## Running Tests

To run the test suite first invoke the following command within the repo, installing the development dependencies:

```bash
$ npm install
```

then run the tests:

```bash
$ npm test
```

## Authors

```bash
$ git summary

 project  : top
 repo age : 1 year, 8 months
 active   : 11 days
 commits  : 27
 files    : 15
 authors  :
    22  fengmk2                 81.5%
     4  Jackson Tian            14.8%
     1  coolme200               3.7%
```

## License

(The MIT License)

Copyright (c) 2011 - 2013 fengmk2 &lt;fengmk2@gmail.com&gt;

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.2.3 / 2013-11-13 
==================

  * fix agentkeepalive

0.2.2 / 2013-10-25 
==================

  * Fix message send issue
  * 分离核心方法和API方法到两个文件

0.2.1 / 2013-10-22 
==================

  * 添加千牛消息推送接口 (@JacksonTian)

0.2.0 / 2013-10-07 
==================

  * fixed #8 change long id from Number to String temporary.
  * improve test case

0.1.1 / 2013-09-23 
==================

  * fixed #5 Insufficient isv permissions error

0.1.0 / 2013-09-22 
==================

  * Update version for travis
  * 改进检查参数值的判断
  * Don't throw Exception

0.0.4 / 2012-11-30 
==================

  * support all user apis #4
  * add coveage result
  * add usage


test/client.test.js
/*!
 * top - test/client.test.js
 * Copyright(c) 2012 - 2013 fengmk2 <fengmk2@gmail.com>
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var should = require('should');
var mm = require('mm');
var urllib = require('urllib');
var fs = require('fs');
var path = require('path');
var top = require('../');

var REST_URL = 'http://gw.api.tbsandbox.com/router/rest';

describe('client.test.js', function () {
  var client = top.createClient({
    appkey: '1021178166',
    appsecret: 'sandboxbc0042b2231100842349ad492',
    REST_URL: REST_URL
  });

  afterEach(mm.restore);

  describe('new Client()', function () {
    it('should return Client', function () {
      var client = new top.Client({appkey: 'key', appsecret: 'secret'});
      client.should.be.instanceof(top.Client);
      var clientWithoutNew = top.Client({appkey: 'key', appsecret: 'secret'});
      clientWithoutNew.should.be.instanceof(top.Client);
    });

    it('should throw error when miss appkey or appsecret', function () {
      (function () {
        top.createClient();
      }).should.throw('appkey or appsecret need!');
      (function () {
        top.createClient({ appkey: 'test' });
      }).should.throw('appkey or appsecret need!');
      (function () {
        top.createClient({ appsecret: 'test' });
      }).should.throw('appkey or appsecret need!');
    });
  });

  describe('sign() http://open.taobao.com/doc/detail.htm?id=111#s6', function () {
    it('should equal 990FD28323F67A1EEC29336EDF373C0E', function () {
      var c = top.createClient({
        appkey: 'test',
        appsecret: 'test'
      });
      var params = {
        method: 'taobao.user.get',
        timestamp: '2011-07-01 13: 52:03',
        format: 'xml',
        app_key: 'test',
        v: '2.0',
        fields: 'nick,location.state,location.city',
        nick: '商家测试帐号17',
        sign_method: 'md5',
      };
      c.sign(params).should.equal('990FD28323F67A1EEC29336EDF373C0E');
    });
  });

  describe('timestamp()', function () {
    it('should return yyyy-MM-dd HH:mm:ss format', function () {
      var s = client.timestamp();
      s.should.match(/\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}:\d{2}/);
    });
  });

  describe('request()', function () {
    it('should request success', function (done) {
      client.request({
        method: 'taobao.user.get',
        fields: 'nick,seller_credit',
        nick: 'sandbox_c_1'
      }, function (err, result) {
        should.not.exist(err);
        var user = result.user_get_response.user;
        user.should.have.keys(['nick', 'seller_credit']);
        user.nick.should.equal('sandbox_c_1');
        done();
      });
    });

    it('should throw error when method miss', function (done) {
      client.request({}, function (err) {
        should.exist(err);
        err.message.should.equal('`method` required');
        done();
      });
    });

    it('should return error when method wrong', function (done) {
      client.request({ method: 'not_exists' }, function (err, data) {
        should.exist(err);
        err.message.should.equal('Invalid method, code 22');
        err.code.should.equal(22);
        err.data.should.equal('{"error_response":{"code":22,"msg":"Invalid method"}}');
        done();
      });
    });
  });

  describe('_wrapJSON()', function () {
    it('should convert long id number to string', function () {
      var s = fs.readFileSync(path.join(__dirname, 'fixtures', 'tmc_messages_consume_response.json')).toString();
      s = client._wrapJSON(s);
      var o = JSON.parse(s);
      // console.log('%s, %j', s, o.tmc_messages_consume_response.messages)
      o.tmc_messages_consume_response.messages.tmc_message[0].id.should.equal('7104300007405429232');
    });
  });

  describe('invoke()', function () {
    var _request = urllib.request;
    afterEach(function () {
      urllib.request = _request;
    });

    it('should mock urllib.request() error', function (done) {
      urllib.request = function (url, options, callback) {
        process.nextTick(function () {
          callback(new Error('mock error'));
        });
      };
      client.invoke('taobao.shop.get', {nick: 'abc', fields: '123'}, [], null, 'GET',
      function (err, item) {
        should.exist(err);
        should.not.exist(item);
        err.should.have.property('message', 'mock error');
        done();
      });
    });

    it('should mock urllib.request() json parse error', function (done) {
      urllib.request = function (url, options, callback) {
        process.nextTick(function () {
          callback(null, '{');
        });
      };
      client.invoke('taobao.shop.get', {nick: 'abc', fields: '123'}, [], null, 'GET',
      function (err, item) {
        should.exist(err);
        should.not.exist(item);
        err.should.have.property('message', 'Unexpected end of input');
        err.should.have.property('name', 'SyntaxError');
        done();
      });
    });
  });

  describe('taobao_user_buyer_get()', function () {
    var _request = urllib.request;
    afterEach(function () {
      urllib.request = _request;
    });
    it('should return buyer', function (done) {
      urllib.request = function (url, options, callback) {
        process.nextTick(function () {
          callback(null, JSON.stringify(require('./fixtures/user_buyer_get_response.json')));
        });
      };
      client.taobao_user_buyer_get({
        session: 'mock',
        fields: 'nick,sex,buyer_credit,avatar,has_shop,vip_info', 
        nick: 'sandbox_c_1'
      }, function (err, user) {
        should.not.exist(err);
        user.should.have.keys('nick,sex,buyer_credit,avatar,has_shop,vip_info'.split(','));
        user.nick.should.equal('sandbox_c_1');
        done();
      });
    });

    it('should return null when session wrong', function (done) {
      client.taobao_user_buyer_get({session: 'mock', fields: 'sex,nick', nick: 'alipublic01notexists'}, 
      function (err, user) {
        should.exist(err);
        err.name.should.equal('TOPClientError');
        err.message.should.equal('Invalid session, code 27; sessionkey-not-generated-by-server: RealSession don&apos;t create by top ! appKey is : 1021178166');
        err.code.should.equal(27);
        err.sub_code.should.equal('sessionkey-not-generated-by-server');
        should.not.exist(user);
        done();
      });
    });

    it('should throw error when nick miss', function (done) {
      client.taobao_user_buyer_get({fields: 'sex,nick'}, function (err, user) {
        should.exist(err);
        err.message.should.equal('`session` required');
        done();
      });
    });
  });

  describe('taobao_user_seller_get()', function () {
    var _request = urllib.request;
    afterEach(function () {
      urllib.request = _request;
    });
    it('should return seller', function (done) {
      var fields = 'user_id,nick,sex,seller_credit,type,has_more_pic,item_img_num,item_img_size,prop_img_num,prop_img_size,auto_repost,promoted_type,status,alipay_bind,consumer_protection,avatar,liangpin,sign_food_seller_promise,has_shop,is_lightning_consignment,has_sub_stock,is_golden_seller,vip_info,magazine_subscribe,vertical_market,online_gaming';
      urllib.request = function (url, options, callback) {
        process.nextTick(function () {
          var user = require('./fixtures/user_seller_get_response.json');
          var keys = fields.split(',');
          var seller = {user_seller_get_response: {user: {}}};
          for (var i = 0; i < keys.length; i++) {
            var k = keys[i];
            seller.user_seller_get_response.user[k] = user.user_seller_get_response.user[k];
          }
          callback(null, JSON.stringify(seller));
        });
      };
      client.taobao_user_seller_get({
        session: 'mock',
        fields: fields, 
        nick: 'hz0799'
      }, function (err, user) {
        should.not.exist(err);
        user.should.have.keys(fields.split(','));
        user.nick.should.equal('hz0799');
        done();
      });
    });

    it('should return err when session wrong', function (done) {
      client.taobao_user_seller_get({session: 'mock', fields: 'sex,nick', nick: 'alipublic01notexists'}, 
      function (err, user) {
        should.exist(err);
        err.name.should.equal('TOPClientError');
        err.message.should.equal('Invalid session, code 27; sessionkey-not-generated-by-server: RealSession don&apos;t create by top ! appKey is : 1021178166');
        should.not.exist(user);
        done();
      });
    });

    it('should throw error when nick miss', function (done) {
      client.taobao_user_seller_get({fields: 'sex,nick'}, function (err, user) {
        should.exist(err);
        err.message.should.equal('`session` required');
        done();
      });
    });
  });

  describe('taobao_user_get()', function () {
    it('should return user', function (done) {
      client.taobao_user_get({fields: 'seller_credit,nick', nick: 'sandbox_c_1'}, 
      function (err, user) {
        should.not.exist(err);
        user.should.have.keys(['seller_credit', 'nick']);
        user.nick.should.equal('sandbox_c_1');
        done();
      });
    });

    it('should return null when nick not exists', function (done) {
      client.taobao_user_get({fields: 'user_id,nick', nick: 'alipublic01notexists'}, 
      function (err, user) {
        should.not.exist(err);
        should.not.exist(user);
        done();
      });
    });

    it('should throw error when nick miss', function (done) {
      client.taobao_user_get({fields: 'user_id,nick'}, function (err, user) {
        should.exist(err);
        err.message.should.equal('`nick` required');
        done();
      });
    });
  });

  describe('taobao_users_get()', function () {
    it('should return users list', function (done) {
      client.taobao_users_get({ fields: 'seller_credit,nick', nicks: 'sandbox_c_2,sandbox_c_1' },
      function (err, users) {
        should.not.exist(err);
        users.should.length(2);
        for (var i = users.length; i--;) {
          var user = users[i];
          user.should.have.keys(['seller_credit', 'nick']);
          user.nick.should.equal(i === 0 ? 'sandbox_c_2' : 'sandbox_c_1');
        }
        done();
      });
    });

    it('should return 4 length list when nicks are all same', function (done) {
      client.taobao_users_get({ 
        fields: 'nick', 
        nicks: 'sandbox_c_1,sandbox_c_2,sandbox_c_3,sandbox_c_4'
      }, function (err, users) {
        should.not.exist(err);
        users.should.length(4);
        for (var i = users.length; i--;) {
          var user = users[i];
          user.should.have.keys(['nick']);
          user.nick.should.match(/sandbox_c_\d/);
        }
        done();
      });
    });

    it('should return 1 length list when one nick not exists', function (done) {
      client.taobao_users_get({fields: 'nick', nicks: 'sandbox_c_3,苏千notexists'},
      function (err, users) {
        should.not.exist(err);
        users.should.length(1);
        for (var i = users.length; i--;) {
          var user = users[i];
          user.should.have.keys(['nick']);
          user.nick.should.equal('sandbox_c_3');
        }
        done();
      });
    });

    it('should return [] no nick exists', function (done) {
      client.taobao_users_get({ fields: 'user_id,nick', nicks: '苏千苏千notexists2,苏千notexists' }, 
      function (err, users) {
        should.not.exist(err);
        users.should.length(0);
        done();
      });
    });

    it('should throw error when nicks miss', function () {
      client.taobao_users_get({ fields: 'user_id,nick' }, function (err, user) {
        should.exist(err);
        err.message.should.be.equal('`nicks` required');
      });
    });
  });

  describe('tmall_selected_items_search()', function () {
    var mockData = JSON.stringify({
      "tmall_selected_items_search_response": {
        "item_list": {
          "selected_item": [{
            "cid": 1101,
            "num_iid": 13088700250,
            "shop_id": 59227746,
            "item_score": "67.33659988217163"
          }]
        }
      }
    });
    var _request = urllib.request;
    after(function () {
      urllib.request = _request;
    });
    // api permission required
    it('should return items', function (done) {
      urllib.request = function (url, options, callback) {
        process.nextTick(function () {
          callback(null, mockData);
        });
      };
      client.tmall_selected_items_search({cid: 50016349}, function (err, items) {
        should.not.exist(err);
        should.exist(items);
        items.should.be.an.instanceof(Array).with.length(1);
        items[0].should.have.keys('cid', 'num_iid', 'shop_id', 'item_score');
        client.taobao_item_get({
          num_iid: items[0].num_iid,
          fields: 'item_img.url,title,price'
        }, done);
      });
    });

    it('should return parameter missing error', function (done) {
      client.tmall_selected_items_search({}, function (err, items) {
        should.exist(err);
        err.name.should.equal('ParameterMissingError');
        err.message.should.equal('`cid` required');
        done();
      });
    });
  });

  describe('taobao_jindoucloud_message_send()', function () {
    var mockData = JSON.stringify({
      "jindoucloud_message_send_response": {
        "send_results": {
          "send_result": [{
            "nick": "nick",
            "err_msg": "nick is null",
            "err_code": "isv.invalid-parameter",
            "is_success": "false"
          }]
        }
      }
    });

    before(function () {
      mm.data(urllib, 'request', mockData);
    });

    after(function () {
      mm.restore();
    });

    // api permission required
    it('should return items', function (done) {
      var params = {
        messages: [{
          "nick": "nick",
          "title": "title",
          "view_data": ["a", "b"],
          "biz_data": {"k1": "v1", "k2": "v2"},
          "biz_id": 0,
          "send_no": 0,
          "msg_category": "item",
          "msg_type": "ItemCreate"
        }]
      };
      client.taobao_jindoucloud_message_send(params, function (err, items) {
        should.not.exist(err);
        should.exist(items);
        items.should.be.an.instanceof(Array).with.length(1);
        items[0].should.have.keys('nick', 'err_msg', 'err_code', 'is_success');
        done();
      });
    });

    it('should return parameter missing error', function (done) {
      var params = {};
      client.taobao_jindoucloud_message_send(params, function (err, items) {
        should.exist(err);
        err.name.should.equal('ParameterMissingError');
        err.message.should.equal('`messages` required');
        done();
      });
    });
  });

  describe('taobao_shop_get()', function () {
    it('should return shop info', function (done) {
      client.taobao_shop_get({nick: 'sandbox_c_1', fields: 'sid,cid,title,nick,desc,bulletin,pic_path,created,modified'},
      function (err, item) {
        should.not.exist(err);
        item.should.have.keys('sid,cid,title,nick,desc,bulletin,pic_path,created,modified'.split(','));
        done();
      });
    });

    it('should return null when user no shop', function (done) {
      client.taobao_shop_get({nick: 'sandbox_c_1_notexists', fields: 'sid,cid,title,nick,desc,bulletin,pic_path,created,modified'},
      function (err, item) {
        should.not.exist(err);
        should.not.exist(item);
        done();
      });
    });

    it('should return parameter missing error', function (done) {
      client.taobao_shop_get({}, function (err, items) {
        should.exist(err);
        err.name.should.equal('ParameterMissingError');
        err.message.should.equal('`nick` required');
        done();
      });
    });

    it('should return mock error', function (done) {
      var errRes = {
        error_response: {
          code: 11,
          msg: 'Insufficient isv permissions',
          sub_code: 'isv.permission-ip-whitelist-limit',
          sub_msg: 'The appkey 4272 is only allowed to call from 10.232.*.*,10.13.*.*,10.62.*.*,10.20.*.*,10.235.*.*,110.75.*.*,10.32.*.*,10.14.*.*,10.12.*.*,10.250.*.*,10.253.*.*,10.17.*.*,10.9.*.*,10.7.*.*,10.33.*.*,10.5.*.*,10.1.*.*,127.0.*.*,10.209.*.*,10.208.*.*,10.249.*.*,10.8.*.*, but your ip is 10.68.179.114'
        }
      };
      mm.http.request(/\//, JSON.stringify(errRes));
      client.taobao_shop_get({nick: 'sandbox_c_1', fields: 'sid,cid,title,nick,desc,bulletin,pic_path,created,modified'},
      function (err, item) {
        should.exist(err);
        err.name.should.equal('TOPClientError');
        err.message.should.include('Insufficient isv permissions, code 11; isv.permission-ip-whitelist-limit: ');
        err.code.should.equal(11);
        err.sub_code.should.equal('isv.permission-ip-whitelist-limit');
        should.not.exists(item);
        done();
      });
    });
  });
});


lib/client.js
/*!
 * top - lib/client.js
 * Copyright(c) 2012 - 2013 fengmk2 <fengmk2@gmail.com> (http://fengmk2.github.com)
 * MIT Licensed
 */

/**
 * Module dependencies.
 */

var utility = require('utility');
var urllib = require('urllib');
var util = require('./util');
var Agent = require('agentkeepalive');
var HttpsAgent = Agent.HttpsAgent;

/**
 * API Client.
 * 
 * @param {Object} options, must set `appkey` and `appsecret`.
 * @constructor
 */
function Client(options) {
  if (!(this instanceof Client)) {
    return new Client(options);
  }
  options = options || {};
  if (!options.appkey || !options.appsecret) {
    throw new Error('appkey or appsecret need!');
  }
  this.REST_URL = options.REST_URL || 'http://gw.api.taobao.com/router/rest';
  this.appkey = options.appkey;
  this.appsecret = options.appsecret;
  this.isHttps = this.REST_URL.indexOf('https://') === 0;
  if (this.isHttps) {
    this.agent = new HttpsAgent();
  } else {
    this.agent = new Agent();
  }
  this.agent.maxSockets = 100;
}

/**
 * Invoke an api by method name.
 * 
 * @param {String} method, method name
 * @param {Object} params
 * @param {Array} reponseNames, e.g. ['tmall_selected_items_search_response', 'tem_list', 'selected_item']
 * @param {Object} defaultResponse
 * @param {String} type
 * @param {Function(err, response)} callback
 */
Client.prototype.invoke = function (method, params, reponseNames, defaultResponse, type, callback) {
  params.method = method;
  this.request(params, type, function (err, result) {
    if (err) {
      return callback(err);
    }
    var response = result;
    if (reponseNames && reponseNames.length > 0) {
      for (var i = 0; i < reponseNames.length; i++) {
        var name = reponseNames[i];
        response = response[name];
        if (response === undefined) {
          break;
        }
      }
    }
    if (response === undefined) {
      response = defaultResponse;
    }
    callback(null, response);
  });
};

Client.prototype._wrapJSON = function (s) {
  // warp fix #8: https://github.com/fengmk2/top/issues/8
  var matchs = s.match(/\"id\"\:\s?\d{16,}/g);
  if (matchs) {
    for (var i = 0; i < matchs.length; i++) {
      var m = matchs[i];
      s = s.replace(m, '"id":"' + m.split(':')[1].trim() + '"');
    }
  }
  return s;
};

var IGNORE_ERROR_CODES = {
  'isv.user-not-exist:invalid-nick': 1,
  // 'isv.user-not-exist:invalid-nick': 1,
};

/**
 * Request API.
 * 
 * @param {Object} params
 * @param {String} [type='GET']
 * @param {Function(err, result)} callback
 * @public
 */
Client.prototype.request = function (params, type, callback) {
  if (typeof type === 'function') {
    callback = type;
    type = null;
  }
  var err = util.checkRequired(params, 'method');
  if (err) {
    return callback(err);
  }
  var args = {
    timestamp: this.timestamp(),
    format: 'json',
    app_key: this.appkey,
    v: '2.0',
    sign_method: 'md5'
  };
  for (var k in params) {
    args[k] = params[k];
  }
  args.sign = this.sign(args);
  type = type || 'GET';
  var options = {type: type, data: args, agent: this.agent};
  var that = this;
  urllib.request(that.REST_URL, options, function (err, buffer) {
    var data;
    if (buffer) {
      buffer = that._wrapJSON(buffer.toString());
      try {
        data = JSON.parse(buffer);
      } catch (e) {
        err = e;
        e.data = buffer.toString();
        data = null;
      }
    }
    var errRes = data && data.error_response;
    if (errRes) {
      if (!errRes.sub_msg || !IGNORE_ERROR_CODES[errRes.sub_code]) {
        // no sub_msg error, let caller handle it.
        // request() only handle root error, like code 40, 41 error.
        var msg = errRes.msg + ', code ' + errRes.code;
        if (errRes.sub_msg) {
          msg += '; ' + errRes.sub_code + ': ' + errRes.sub_msg;
        }
        err = new Error(msg);
        err.name = 'TOPClientError';
        err.code = errRes.code;
        err.sub_code = errRes.sub_code;
        err.data = buffer.toString();
        data = null;
      }
    }

    callback(err, data);
  });
};

/**
 * Get now timestamp with 'yyyy-MM-dd HH:mm:ss' format.
 * @return {String}
 */
Client.prototype.timestamp = function () {
  return utility.YYYYMMDDHHmmss();
};

/**
 * Sign API request.
 * see http://open.taobao.com/doc/detail.htm?id=111#s6
 * 
 * @param  {Object} params
 * @return {String} sign string
 */
Client.prototype.sign = function (params) {
  var sorted = Object.keys(params).sort();
  var basestring = this.appsecret;
  // (md5(secretkey1value1key2value2...secret))
  for (var i = 0, l = sorted.length; i < l; i++) {
    var k = sorted[i];
    basestring += k + params[k];
  }
  basestring += this.appsecret;
  return utility.md5(basestring).toUpperCase();
};

var api = require('./api');
// 扩展原型方法
Object.keys(api).forEach(function (key) {
  Client.prototype[key] = api[key];
});

exports.Client = Client;


lib/api.js
var util = require('./util');

/**
 * Get taobao user by nick.
 * 
 * @see http://api.taobao.com/apidoc/api.htm?path=cid:1-apiId:1
 * @param {Object} params, `nick` or `session` must set one.
 *  - {String} fields, (required)fields of user, e.g.: 'user_id,nick,seller_credit'.
 *  - {String} nick, (optional)user nick name.
 *  - {String} session, (optional)session id.
 * @param {Function(err, user)} callback
 * @public
 */
exports.taobao_user_get = function (params, callback) {
  var keys = ['fields'];
  if (!params.session) {
    keys.push('nick');
  }
  var err = util.checkRequired(params, keys);
  if (err) {
    return callback(err);
  }
  this.invoke('taobao.user.get', params, ['user_get_response', 'user'], null, 'GET', callback);
};

/**
 * Get taobao current buyer.
 * 
 * @see http://api.taobao.com/apidoc/api.htm?path=cid:1-apiId:21348
 * @param {Object} params, `nick` or `session` must set one.
 *  - {String} fields, (required)fields of user, e.g.: 'nick,sex'.
 *  - {String} access_token, (optional) oauth token, access_token or session must set one.
 *  - {String} session, (optional)session id.
 * @param {Function(err, user)} callback
 * @public
 */
exports.taobao_user_buyer_get = function (params, callback) {
  var keys = ['fields'];
  if (!params.access_token) {
    keys.push('session');
  }
  var err = util.checkRequired(params, keys);
  if (err) {
    return callback(err);
  }
  this.invoke('taobao.user.buyer.get', params,
    ['user_buyer_get_response', 'user'], null, 'GET', callback);
};

/**
 * Get taobao users by nick list.
 * 
 * @param  {Object} params
 *  - {String} fields, (required)fields of user. 
 *  - {String} nicks, (required)user nick name list, separate by (,), e.g.: mk2,aerdeng
 * @param  {Function(err, users)} callback
 * @public
 */
exports.taobao_users_get = function (params, callback) {
  var err = util.checkRequired(params, [ 'fields', 'nicks' ]);
  if (err) {
    return callback(err);
  }
  this.invoke('taobao.users.get', params,
    ['users_get_response', 'users', 'user'], [], 'GET', callback);
};

/**
 * Get taobao current seller.
 * 
 * @see http://api.taobao.com/apidoc/api.htm?path=cid:1-apiId:21349
 * @param {Object} params, `nick` or `session` must set one.
 *  - {String} fields, (required)fields of user, e.g.: 'nick,sex'.
 *  - {String} access_token, (optional) oauth token, access_token or session must set one.
 *  - {String} session, (optional)session id.
 * @param {Function(err, user)} callback
 * @public
 */
exports.taobao_user_seller_get = function (params, callback) {
  var keys = ['fields'];
  if (!params.access_token) {
    keys.push('session');
  }
  var err = util.checkRequired(params, keys);
  if (err) {
    return callback(err);
  }
  this.invoke('taobao.user.seller.get', params,
    ['user_seller_get_response', 'user'], null, 'GET', callback);
};

/**
 * tmall.selected.items.search 天猫类目精选商品库
 * 
 * @see http://api.taobao.com/apidoc/api.htm?path=cid:10240-apiId:11116
 * @param {Object} params
 *  - {Number|String} cid
 * @param {Function(err, items)} callback
 *  - {Array} items [{
 *    cid: 1101,
 *    num_iid: 13088700250,
 *    shop_id: 59227746,
 *    item_score: "67.33659988217163"
 *  }, ...]
 * @public
 */
exports.tmall_selected_items_search = function (params, callback) {
  var err = util.checkRequired(params, ['cid']);
  if (err) {
    return callback(err);
  }
  this.invoke('tmall.selected.items.search', params,
    ['tmall_selected_items_search_response', 'item_list', 'selected_item'], [],
    'GET', callback);
};

exports.taobao_item_get = function (params, callback) {
  var err = util.checkRequired(params, ['num_iid', 'fields']);
  if (err) {
    return callback(err);
  }
  this.invoke('taobao.item.get', params,
    ['item_get_response', 'item'], {}, 'GET', callback);
};

/**
 * Get taobao shop by nick.
 *
 * @see http://api.taobao.com/apidoc/api.htm?spm=0.0.0.34.b8913b&path=cid:9-apiId:68
 * @param  {Object} params
 *  - {String} fields, (required)fields of user.
 *  - {String} nick, (optional)user nick name.
 *  - {String} session, (optional)session id.
 * @param  {Function(err, user)} callback
 * @public
 */
exports.taobao_shop_get = function (params, callback) {
  var err = util.checkRequired(params, ['nick', 'fields']);
  if (err) {
    return callback(err);
  }
  this.invoke('taobao.shop.get', params,
    ['shop_get_response', 'shop'], null, 'GET', callback);
};

/**
 * Send qianniu message.
 *
 * @see http://api.taobao.com/apidoc/api.htm?path=cid:20378-apiId:22542
 * @param {Object} params
 *  - {Array} messages, (required)Message list.
 * @param  {Function(err, result)} callback
 * @public
 */
exports.taobao_jindoucloud_message_send = function (params, callback) {
  var err = util.checkRequired(params, ['messages']);
  if (err) {
    return callback(err);
  }
  this.invoke('taobao.jindoucloud.message.send', params,
    ['jindoucloud_message_send_response', 'send_results', 'send_result'], null, 'GET', callback);
};
